### ISSUE TITLE
useEffect with async function call causes `act(...)` warning

### ISSUE THREAD

**Issue by fabb** [testing-library/react-testing-library#667]
When rendering and testing a component with a `useEffect` that calls an async function that modifies the component's state, then an `act()` console error will be output when running tests.

- `@testing-library/react` version: 10.0.4
- `jest` version: 26.0.1
- DOM Environment: `jsdom` version: 16.2.2

### Relevant code or config:

```js
const MyComponent = () => {
    const [loaded, setLoaded] = useState(false)

    useEffect(() => {
        async function load() {
            await loadFromApi()
            setLoaded(true)
        }
        load()
    }, [])

    return loaded ? <div>loaded</div> : <div>loading...</div>
}

// timeout=0 to simulate mocked api responses
const loadFromApi = () => new Promise(resolve => setTimeout(resolve, 0))
```

### What you did:

Just render this component in a jest test.

```js
test('act() console error', () => {
    const sut = render(<MyComponent />)
    expect(sut).toBeDefined()
})
```

### What happened:

This error is output in my project when running tests:

```
  console.error
    Warning: An update to null inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
```

### Reproduction:

See console output:
https://codesandbox.io/s/react-testing-library-demo-sk2so

### Problem description:

Rendering should wait until everything has been resolved.

### Suggested solution:

Since async functions inside `useEffect` are quite common, rendering a component containing such a hook should wait for the component updates to have finished (with a small timeout).

**Comment by kentcdodds** [comment]
Hi @fabb,

There's no way for React Testing Library to know that you've got async stuff happening in the background, and you wouldn't want that anyway because you probably want to assert the "loading" state anyway.

This is why React Testing Library gives you [async utils](https://testing-library.com/docs/dom-testing-library/api-async) which you can use to wait for the UI to update asynchronously.

Here's how you'd fix that codesandbox: https://codesandbox.io/s/react-testing-library-demo-dmklb?file=/src/__tests__/hello.js:244-306

The cool thing about this is the async utils provided by React Testing Library are automatically wrapped in `act` so you shouldn't ever have to worry about using `act` manually. Learn more here: https://kcd.im/act-warning

Good luck :)

**Comment by fabb** [comment]
Thanks for looking into this. I guess I need to find out what I need to wait for. Let‘s go hunting 😅

**Comment by fabb** [comment]
BTW thanks for the tip of updating jest to include a backtrace in the error log output, that helps a lot.

**Comment by CarlosIUSalazar** [comment]
> Hi @fabb,
> 
> There's no way for React Testing Library to know that you've got async stuff happening in the background, and you wouldn't want that anyway because you probably want to assert the "loading" state anyway.
> 
> This is why React Testing Library gives you [async utils](https://testing-library.com/docs/dom-testing-library/api-async) which you can use to wait for the UI to update asynchronously.
> 
> Here's how you'd fix that codesandbox: https://codesandbox.io/s/react-testing-library-demo-dmklb?file=/src/__tests__/hello.js:244-306
> 
> The cool thing about this is the async utils provided by React Testing Library are automatically wrapped in `act` so you shouldn't ever have to worry about using `act` manually. Learn more here: https://kcd.im/act-warning
> 
> Good luck :)

OMG I was stuck with this warning for 24 hours.  Thank you for your help, this worked.

**Comment by kopax** [comment]
Hi guys, I am having the following test wrapped in an async function: 

```
    const renderAPI = await render(<PrivacyPolicy />)
```

This throws warning:

```jsx
 console.error
    Warning: An update to PrivacyPolicy inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
    
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */
    
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
        in PrivacyPolicy

       8 | 
       9 |   useEffect(() => {
    > 10 |     storage.readObject('has_accepted_cookie').then((value) => setShouldShowModal(!value))
         |                                                               ^
      11 |   }, [])
```

None of the advice here helped, any clue?

**Comment by carlnc** [comment]
1. Turn on fake timers

```typescript
        beforeEach(() => {
            jest.useFakeTimers();
        });
```

2. Mock React.useState with a helper method - defer based on setTimeout()

```typescript
        function deferUseState(): () => void {
            const { useState } = React;
            React.useState = (<T extends unknown>(defaultVal: T): [T, Dispatch<T>] => {
                const [value, updateValue] = useState<T>(defaultVal);

                const wrappedUpdate = (newValue: T) => {
                    setTimeout(() => { updateValue(newValue); }, 1);
                };

                return [value, wrappedUpdate];
            }) as unknown as typeof React.useState;

            return () => { React.useState = useState; }; // restore function
        }
```

3. Update test to use helper method.

```typescript
const restoreUseState = deferUseState();
// do render as normal
restoreUseState();

// now call act/runTimers as required.
act(() => {
    jest.runAllTimers();
});

**Comment by kevinvugts** [comment]
Hi @kentcdodds,

Considering I am borrowing one of your lovely hooks called useAsync() in combination with the AuthProvider context provider to either render an authenticated or unauthenticated app. How would we be able to test the login functionality for example?

**AuthContext / making use of the useAsync hook:** https://pastebin.com/9g5ZNe15 

**What would I like to test?**
I would like to be able to test if the user is actually logged in and redirected to, in this case, a dashboard page containing a text.

**I have the following test suite so far:** https://pastebin.com/KjaTpvEP 

**What is the problem?**
Because or AuthContext is responsible for setting a user into context and this is happening when triggering the login endpoint, I am wondering how I could replicate this in the test. The code I am using is written by you @kentcdodds in the epic react course and the bookshelf app.

**Comment by kentcdodds** [comment]
You could bring this question to my https://kcd.im/office-hours

Can't answer here/now

**Comment by kevinvugts** [comment]
Awesome! Definitely will raise this question next office meeting.
Have put the question inside Discord.

I think it's quite an interesting / complex one!

**Comment by kevinvugts** [comment]
> You could bring this question to my https://kcd.im/office-hours
> 
> Can't answer here/now

Hi @kentcdodds ,

Thanks for your reply in the discord channel. It really helpt me along the way!
Right now I am looking into the bookshelf tests and basically copying some parts of it.

I am struggling with 1 part in the test process, and that has all to do with the   waitForElementToBeRemoved function.
For some reason I am getting a time error over and over again, no matter which query I use. (getBy, findBy or the preferred one queryBy)

Here is the test I am writing:

```
async function renderDashboardScreen({ user } = {}) {
  if (user === undefined) {
    user = await loginAsUser(); <-- Similar function you wrote in the [bookshelf](https://github.com/kentcdodds/bookshelf/blob/main/src/test/app-test-utils.js#L31).
  }

  const route = `/dashboard/`;

  const utils = await render(<App />, { user, route });

  return {
    ...utils,
    user,
  };
}

test("renders the dashboard screen and is checking if string is present after fully loaded", async () => {
  await renderDashboardScreen();

  await waitForElementToBeRemoved(() => screen.getAllByText("Laden..."), {
    timeout: 10000,
  }); <--- It does not get past this point. The functionality I am testing is the same as the list of books in the bookshelf app.

  expect(await screen.getByText(/Tom Dirven/i));
}
```

**Question**
What could possibly go wrong in this case? See the attached video for a visual reference what is happening in terms of state updates. Initial state is a skeleton state from react query, after executing the actual api request the data is updated accordingly. I am also using the exact same [api-client.js](https://github.com/kentcdodds/bookshelf/blob/main/src/utils/api-client.js) as you use in the bookshelf repo.

I am looking forward for your answer! Thanks in advance for any effort.

**Comment by kentcdodds** [comment]
@kevinvugts, could you bring this to https://testing-library.com/discord or https://kentcdodds.com/discord instead?

**Comment by queeniema** [comment]
@kentcdodds In the codesandbox you provided, when I modify the `setTimeout` to a longer period (10 sec), the test fails (see [here](https://codesandbox.io/s/react-testing-library-demo-forked-n0b0d?file=/src/__tests__/hello.js)).

```js
const loadFromApi = () => new Promise((resolve) => setTimeout(resolve, 10000));
```

It doesn't seem to be waiting for the promise to resolve (total test time is ~ 4.5 sec). Any suggestions on how to get this to work for long-running tests?

**Comment by kentcdodds** [comment]
@queeniema, I don't typically write jest tests that hit the real API and take a long time to run. I mock things like that so they're fast and then test those actual calls with an E2E testing tool like Cypress.

If you have any follow-up questions, please take them to https://testing-library.com/discord

**Comment by MatanBobi** [comment]
> @kentcdodds In the codesandbox you provided, when I modify the `setTimeout` to a longer period (10 sec), the test fails (see [here](https://codesandbox.io/s/react-testing-library-demo-forked-n0b0d?file=/src/__tests__/hello.js)).
> 
> ```js
> const loadFromApi = () => new Promise((resolve) => setTimeout(resolve, 10000));
> ```
> 
> It doesn't seem to be waiting for the promise to resolve (total test time is ~ 4.5 sec). Any suggestions on how to get this to work for long-running tests?

@queeniema 
The reason it's failing is because `screen.findByText(/loaded/i)` uses `waitFor` and the default timeout is `1000ms`. You'll need to change it to a higher number for the `findBy*` query to wait longer. You can see the API [here](https://testing-library.com/docs/dom-testing-library/api-async/#findby-queries)

**Comment by ChiralMichael** [comment]
@kentcdodds But why does the testing library NOT cause my clean-up function to be called? This should be part of any proper teardown of the component. I am following best practices in my useEffects to not do async updates when my component is torn down. Now my tests need to concern themselves with things that are not germane to the test. There are many tests that don't care about the async operation and I shouldn't have to wait for it.

**Comment by ChiralMichael** [comment]
Additional information: If I explicitly [unmount](https://testing-library.com/docs/react-testing-library/api/#unmount) my components I can cause the useffect cleanup functions to be invoked. It appears that there is no unmounting during [cleanup](https://testing-library.com/docs/react-testing-library/api#cleanup)? That is unexpected.

**Comment by kentcdodds** [comment]
There *is* unmounting in cleanup, but it's possible jest is exiting the child process before anything has a chance to flush. If you want to test cleanup you need to unmount explicitly. There's not much Testing Library can do about this. Testing Library *definitely* cleans up automatically for you. It's Jest that's exiting the process before the rest of your code gets to run 🤷‍♂️

**Comment by ChiralMichael** [comment]
Thanks for the quick reply :-) I'm not testing cleanup. I'm testing a component that *happens* to have an async operation and also a cleanup function. The assertion of my test doesn't rely on that async op or the cleanup. Testing the component at all has this error because at the time the async operations returns, the cleanup function has yet to run, but apparently the react context has been thrown away. The async implementation here is just a test stub that returns `Promise.resolve(true)` so there's no delay but we are put in the event loop regardless. As we talk about it, if I had to guess, there's probably a race condition here and I guess the cleanup probably needs to happen a tick sooner. I'll see about getting a repro online somewhere for you. Should I open a new issue? I did hijack this one.
